### PR TITLE
Create a more usable folder structure for downloads

### LINKED ISSUE #383: Bulking downloading folder structure
I've noticed that in practice, it's really helpful if each message/document is in it's own folder. The reason is because in Tails, you double-click a .gpg file to decrypt it, and it creates the new file in the same folder. If it's a doc, then you right-click on the .zip to extract it to get the original file that was uploaded. Sometimes (if the source is savvy and follows the journalist key instructions) this is another .gpg file with a different name, and you double-click on that to decrypt it. And sometimes, it decrypts into a .zip or .tar.gz or something that the source sent, so you extract that. Etc, etc.

Basically, each message generates at least 2 files, sometimes 3 (if they post a gpg-encrypted message in the box). And each doc generates at least 3 files (.zip.gpg, .zip, and whatever is extracted from the .zip), sometimes way more. It gets really disorganized to have all of the files just sitting there in the same folder.

(Also, sometimes a source tries to upload the same document twice because they weren't sure if it worked the first time. If the two .zip.gpg files are in the same folder, you end up extracting the zip files over each other when the contents have the same filename.)

So anyway, I propose that the bulk download feature (and any future bulk download features, such as #372) shouldn't just put all the messages/docs into a zip file, but rather should build a directory structure like this:

```
| source_codename
|--| year-month-date
|--|--| doc_number
|--|--|--| document.gpg
```

So for example, if the source codename is "Ottoman mate" and they uploaded a message and then a document on May 22, the directory structure would be:

```
ottoman-mate
ottoman-mate/2014-05-22
ottoman-mate/2014-05-22/1
ottoman-mate/2014-05-22/1/1-ottoman_mate-msg.gpg
ottoman-mate/2014-05-22/2
ottoman-mate/2014-05-22/2/2-ottoman_mate-doc.zip.gpg
```

Make sense? I don't think the directory structure needs to be exactly like this, as long as it's organized sequentially and each message/doc is in it's own folder. (And the dates are super useful, especially if there's weeks of back-and-forth between journalist and source, to look back -- but the date could just be the date of last activity.)

### PR DESCRIPTION
**Pull request by ajvb**
This PR comes based on a conversation with @fowlslegs
and further information can be found here:
https://github.com/freedomofpress/securedrop/issues/383#issuecomment-258656017

Fixes #383

### CODE REVIEW

**Comment by ajvb** [comment]
If https://github.com/freedomofpress/securedrop/pull/1456 is merged before this, then this will need some work to be merged properly. More than happy to do so.

**Comment by fowlslegs** [comment]
Going to wait for #1467 to be merged, then I'll ping you to rebase this.

**Comment by fowlslegs** [comment]
Note #1467 is in turn waiting for #1440. Sorry to make you wait, just the order that seems to make the most sense.

**Comment by ajvb** [comment]
No worries :)

**Comment by fowlslegs** [comment]
Status: #1440 merged, #1467 rebased on develop, and awaiting review (if you want to review just the last commit that involved a rebase, and re-factor of the test suite, that'd be :100:). You should find some convenient new code in `securedrop/tests/common.py` that should help make unit tests better and easier to write.

**Comment by ajvb** [comment]
@fowlslegs Ok awesome! Just reviewed #1467, though it looks like it needs one last rebase? Anyways, as soon as its in ping me and I'll rebase this PR and refactor the test code!

**Comment by fowlslegs** [comment]
Ping!

**Comment by ajvb** [comment]
@fowlslegs Pong!

File: securedrop/store.py [hunk 1]
```diff
@@ -68,17 +68,27 @@ def path(*s):
...
-        for filename in filenames:
-            verify(filename)
-            zip.write(filename, arcname=os.path.join(
-                zip_directory,
-                os.path.basename(filename)
-            ))
+        for source in sources:
+            submissions = [s for s in items_selected if s.source.journalist_designation == source]
+            i = 1
+            for submission in submissions:
```
Comment (conorsch) [inline securedrop/store.py]: :+1:

**Comment by redshiftzero** [comment]
Hey @ajvb just pinging ya on this one - let me know if you need help to get this over the finish line

**Comment by redshiftzero** [comment]
Cherry picked your commits and made that small change in a superceding PR #1486, so closing this one. Thanks for your work on this @ajvb!

**Pull request closed without merging**